common/functions/src/scalars/dates/number_function.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2020 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::marker::PhantomData;
+
+use common_datavalues::arrays::DFUInt32Array;
+use common_datavalues::chrono::DateTime;
+use common_datavalues::chrono::Datelike;
+use common_datavalues::chrono::TimeZone;
+use common_datavalues::chrono::Utc;
+use common_datavalues::prelude::*;
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::scalars::Function;
+
+#[derive(Clone, Debug)]
+pub struct NumberFunction<T> {
+    display_name: String,
+    t: PhantomData<T>,
+}
+
+pub trait NumberResultFunction {
+    fn execute(_value: DateTime<Utc>) -> u32;
+}
+
+#[derive(Clone)]
+pub struct ToYYYYMM;
+
+impl NumberResultFunction for ToYYYYMM {
+    fn execute(value: DateTime<Utc>) -> u32 {
+        value.year() as u32 * 100 + value.month()
+    }
+}
+
+impl<T> NumberFunction<T>
+where T: NumberResultFunction + Clone + Sync + Send + 'static
+{
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+        Ok(Box::new(NumberFunction::<T> {
+            display_name: display_name.to_string(),
+            t: PhantomData,
+        }))
+    }
+}
+
+impl<T> Function for NumberFunction<T>
+where T: NumberResultFunction + Clone + Sync + Send + 'static
+{
+    fn name(&self) -> &str {
+        self.display_name.as_str()
+    }
+
+    fn return_type(&self, _args: &[DataType]) -> Result<DataType> {
+        Ok(DataType::UInt32)
+    }
+
+    fn num_arguments(&self) -> usize {
+        1
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(false)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, _input_rows: usize) -> Result<DataColumn> {
+        if columns.is_empty() {
+            return Result::Err(ErrorCode::UnknownFunction(
+                "Number of arguments for function toYYYYMM doesn't match: passed 0, should be 1 or 2")
+            );
+        }
+
+        let data_type = columns[0].data_type();
+        let number_vec: Vec<Option<u32>> = match data_type {
+            DataType::Date16 => {
+                let number_vec_result = columns[0].column().to_array()?.u16()?.iter().map(|value|{
```

Comment:
Great, but there are also much more spaces to improve this function.

1. Constant path to check, which may help in `select toYYYYMM( 123455 ) from number(xxx)`
2. Put the whole eval function into `execute` function, this can be done in `toYYYYMMDDhhmmss`, but it's return value is u64.

Reply:
> Constant path to check, which may help in select toYYYYMM( 123455 ) from number(xxx)

It is nice to do this, but there are somethings I am not  sure:
1. If I understand correctly , ` toYYYYMM(123455)` is the same as `select toYYYYMM(toDate(123455))` or `toYYYYMM(toDateTime(123455))`, so which one to choose? toDateTime?
2. `Constant` may be  negative value , what should return for this situation?
3. After this, `select toYYYYMM(number) from number(xxx)` will also work. Is it ok?

Reply:
>  is the same as select toYYYYMM(toDate(123455)) or toYYYYMM(toDateTime(123455))

It's related to `DataType::Date16` or `DataType::DateTime` ..

> Constant may be negative value 

Actually, if its logical type is Date16/DateTime32, which physical type is `u16` and `u32`, the constant value will never be negative.

> toYYYYMM(number)  will never work, we just need to make toYYYYMM(Date/DateTime) work, the inner argument is physically constant.

Reply:
> It's related to DataType::Date16 or DataType::DateTime .

How do I make sure it is a DateType::Date16 or a DataType::DateTime when only giving `select toYYYYMM(123455)`?

Reply:
Let's use 

```
let c = columns[0].column().to_array()?.u16()?;
c.apply_cast_numeric(..)
```
instead, it take consideration of the null values.

Reply:
>Let's use
....
instead, it takes consideration of the null values.

Got it.

> Constant path to check, which may help in select toYYYYMM( 123455 ) from number(xxx)

As `DataColumn.to_array()` has coverage the `Constant` case as Series, what did you mean  `Constant path to check`?

Reply:
Constant path can have better performance, you can simply use to_array(). 
But it will build an array with the same number which has poor performance because you need to calculate each row.

Reply:
Thanks, got it.